# Main File - Calculate Distance to Universities
## Import+Use created dictionaries for data to avoid calculating it every time

In [1]:
import csv
import os
import math

In [3]:
uniname = {}
with open('./dictionaries/uniname.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        uniname[row[0]]=row[1]
uniname

{'1545931': 'Wirtschaftsuniversität Wien',
 '1546113': 'Universität für Musik und darstellende Kunst Wien',
 '1546123': 'Universität für angewandte Kunst',
 '1546137': 'Medizinische Universität Wien',
 '1546148': 'Universität für Bodenkultur Wien',
 '1546202': 'Technische Universität Wien',
 '1546253': 'Campus der Universität Wien',
 '1546266': 'Webster University Vienna',
 '1546291': 'Universität Wien Hauptgebäude',
 '1546292': 'Universitätssportzentrum Wien'}

In [81]:
unigps = {}
with open('./dictionaries/unigps.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        unigps[row[0]]=row[1]
#unigps

## Create and Write Unigpsloc Dict [matching UniID and GPS only for selected Unis]

In [5]:
unigpsloc = {}
for k, v in uniname.items():
    if k in unigps.keys():
        unigpsloc[k] = unigps.get(k)
unigpsloc

{'1545931': '48.21384341856702,16.408860446502068',
 '1546113': '48.196989050609844,16.38231295437089',
 '1546123': '48.20838710235376,16.38293943118496',
 '1546137': '48.21963565071546,16.351735473671383',
 '1546148': '48.23654607282594,16.337577480698467',
 '1546202': '48.198976375914846,16.369895225486264',
 '1546253': '48.21624134156441,16.351350792743386',
 '1546266': '48.223429445296574,16.423918403848166',
 '1546291': '48.213012828415096,16.360851928870705',
 '1546292': '48.2054889397581,16.317331757786384'}

In [6]:
# Write Unigpsloc Dict
with open('./dictionaries/unigpsloc.csv', 'w', newline='') as csvfile:
    writer = csv.writer(csvfile, delimiter=',')
    for k, v in unigpsloc.items():
        writer.writerow( [ k, v ] )
unigpsloc

{'1545931': '48.21384341856702,16.408860446502068',
 '1546113': '48.196989050609844,16.38231295437089',
 '1546123': '48.20838710235376,16.38293943118496',
 '1546137': '48.21963565071546,16.351735473671383',
 '1546148': '48.23654607282594,16.337577480698467',
 '1546202': '48.198976375914846,16.369895225486264',
 '1546253': '48.21624134156441,16.351350792743386',
 '1546266': '48.223429445296574,16.423918403848166',
 '1546291': '48.213012828415096,16.360851928870705',
 '1546292': '48.2054889397581,16.317331757786384'}

## Open Car and Bike GPS Files

In [7]:
cargps = {}
with open('./dictionaries/cargps.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        cargps[row[0]]=row[1]
#cargps

In [8]:
bikegps = {}
with open('./dictionaries/bikegps.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        bikegps[row[0]]=row[1]
#bikegps

# Calculate Distances

Code taken from :

## xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

In [9]:
from math import radians, cos, sin, asin, sqrt
def haversine(lon1, lat1, lon2, lat2):
    """
    Calculate the great circle distance between two points 
    on the earth (specified in decimal degrees)
    """
    # convert decimal degrees to radians 
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
    # haversine formula 
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a)) 
    km = 6367 * c
    return km

Create dictionaries for all Unis, which map ID of car/bike with distance to respective uni.

    disto[15446266]
        [4166:511]
        [4223:772]
    disto[1546292]
        

## legacy 
Attempted to create Dictioniaries in a Loop, failed in doing so - cannot create/access dictionaries from function results. 


print(unigpsloc.keys())

for k in unigpsloc.keys():
    str("disto"+k) = {}
    
disto1546266 = {}
disto1546253 = {}
disto1546292 = {}
disto1546291 = {}
disto1546113 = {}
disto1545931 = {}
disto1546123 = {}
disto1546137 = {}
disto1546202 = {}
disto1546148 = {}

# Calculations

## legacy - tried doing it for each one since creation in Loop failed.
'''
disto1545931 = {}
gpsx=float(unigpsloc.get('1545931').split(",")[0])
gpsy=float(unigpsloc.get('1545931').split(",")[1])
for k, v in cargps.items():
    lon2=float(v.split(",")[0])
    lat2=float(v.split(",")[1])
    dis = haversine(gpsx, gpsy, lon2, lat2)
    if dis < 2.0:
        disto1545931[k] = dis
for k, v in bikegps.items():
    lon2=float(v.split(",")[0])
    lat2=float(v.split(",")[1])
    dis = haversine(gpsx, gpsy, lon2, lat2)
    if dis < 2.0:
        disto1545931[k] = dis
    
disto1545931  
'''


## Main
Create a Dictionary in a Dictionaries, calculate Distance from Universities [first dict] to car and bike gps locations, store all Entries below a certain Radius (1km) in the second Dict.

This means the second Dictionary in the University Identifier Dict contains car/bike ID's and distance to Uni (in whichever second Dict it is in)

BikeID always starts with 11, Car with 4 - so we can sort by that afterwards.

In [76]:
disto={}
for unikey in unigpsloc.keys():
    disto[unikey]={} 
    
    gpsx=float(unigpsloc.get(unikey).split(",")[0])
    gpsy=float(unigpsloc.get(unikey).split(",")[1])
    
    for k, v in cargps.items():
        lon2=float(v.split(",")[0])
        lat2=float(v.split(",")[1])
        dis = haversine(gpsx, gpsy, lon2, lat2)
        if dis < 1.0:
            disto[unikey][k] = dis
    for k, v in bikegps.items():
        lon2=float(v.split(",")[0])
        lat2=float(v.split(",")[1])
        dis = haversine(gpsx, gpsy, lon2, lat2)
        if dis < 1.0:
            disto[unikey][k] = dis


In [11]:
for uni in disto.keys():
    print("Plätze: " +str(len(disto[str(uni)]))+ "\t" + uniname[uni])

Plätze: 2	Universität für Bodenkultur Wien
Plätze: 10	Universität für Musik und darstellende Kunst Wien
Plätze: 0	Webster University Vienna
Plätze: 12	Technische Universität Wien
Plätze: 5	Universitätssportzentrum Wien
Plätze: 14	Medizinische Universität Wien
Plätze: 17	Universität für angewandte Kunst
Plätze: 14	Campus der Universität Wien
Plätze: 16	Universität Wien Hauptgebäude
Plätze: 4	Wirtschaftsuniversität Wien


## Adding Statistics
Adding up Car / Bike Totals, Calculating average Distance

st\_ represents statistics_ to filter by afterwards


Webster doesnt have a Bike Parking Space - Got a ZeroDivisionError - hence the try/catch exception.

In [77]:
for unikey in disto.keys():
    car, bike = 0, 0
    carsum, bikesum = 0, 0
    for k, v in disto[str(unikey)].items():
        if k[0] == '4':       #car
            car += 1      
            carsum += float(v)
        if k[0:2] == '11':    #bike
            bike += 1     
            bikesum += float(v)
    
    
    disto[unikey]['st_car']     = str(car)
    disto[unikey]['st_bike']    = str(bike)
    disto[unikey]['st_total']   = str(bike + car)
    try:
        caravg = carsum/car
        disto[unikey]['st_caravg']  = str(caravg)
    except ZeroDivisionError:
        caravg = 0.00
        disto[unikey]['st_caravg']  = 0
    try:
        bikeavg = bikesum/bike
        disto[unikey]['st_bikeavg'] = str(bikeavg)     
    except ZeroDivisionError:
        bikeavg = 0.00
        disto[unikey]['st_bikeavg'] = 0
        
    print("Cars: " + str(car) +"  \t~"+ str(caravg)[0:5]+" km  \tBikes: " + str(bike) +"  \t~"+ str(bikeavg)[0:5] +" km   \t\t"+ uniname[unikey])
    
    

Cars: 1  	~0.918 km  	Bikes: 1  	~0.897 km   		Universität für Bodenkultur Wien
Cars: 3  	~0.574 km  	Bikes: 7  	~0.752 km   		Universität für Musik und darstellende Kunst Wien
Cars: 0  	~0.0 km  	Bikes: 0  	~0.0 km   		Webster University Vienna
Cars: 4  	~0.525 km  	Bikes: 8  	~0.641 km   		Technische Universität Wien
Cars: 2  	~0.864 km  	Bikes: 3  	~0.762 km   		Universitätssportzentrum Wien
Cars: 5  	~0.579 km  	Bikes: 9  	~0.648 km   		Medizinische Universität Wien
Cars: 6  	~0.743 km  	Bikes: 11  	~0.721 km   		Universität für angewandte Kunst
Cars: 8  	~0.758 km  	Bikes: 6  	~0.625 km   		Campus der Universität Wien
Cars: 5  	~0.579 km  	Bikes: 11  	~0.628 km   		Universität Wien Hauptgebäude
Cars: 1  	~0.526 km  	Bikes: 3  	~0.418 km   		Wirtschaftsuniversität Wien


## Store created dictionary [disto]

In [13]:
# Write Unigpsloc Dict
with open('./dictionaries/unidistance.csv', 'w', newline='') as csvfile:
    writer = csv.writer(csvfile, delimiter=',')
    for k, v in disto.items():
        writer.writerow( [ k, v ] )

## Testing Result Dict


In [14]:
for k, v in disto['1546291'].items():
    if k[0:3] == 'st_':
        print(k, v)

st_total 16
st_caravg 0.5799990510384674
st_car 5
st_bike 11
st_bikeavg 0.6286976148934692


In [15]:
for k, v in disto['1546291'].items():      
    if k[0:3] != 'st_':
        print(str(k) +"\t" + str(v))

11258	0.8950766838918074
11208	0.33720125935904655
11263	0.8964277074938065
4204	0.7998247710931291
11201	0.7365912241406238
11214	0.696099345320119
11213	0.11304465707983724
4236	0.4690715208672136
4199	0.7522569658752941
4196	0.12932098549392465
11262	0.8661144789722911
11264	0.925459575136689
11265	0.6747954543083463
11272	0.1779042503445643
4202	0.7495210118627758
11202	0.5969591277810316


In [16]:
# Webster University - No available Spaces within 1km radius

for k, v in disto['1546266'].items():      
    if k[0:3] != 'st_':
        print(str(k) +"\t" + str(v))
        
for k, v in disto['1546266'].items():
    if k[0:3] == 'st_':
        print(k, v)

st_bikeavg 0
st_car 0
st_total 0
st_bike 0
st_caravg 0


# Resolve ID's to Streetnames

## import dictionaries:

bikeidadress.csv

caridadress.csv

In [17]:
bikeidadress = {}
with open('./dictionaries/bikeidadress.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        bikeidadress[row[0]]=row[1]
#bikeidadress

In [18]:
caridadress = {}
with open('./dictionaries/caridadress.csv', 'r') as f:
    file = csv.reader(f, delimiter=",")
    for row in file:
        caridadress[row[0]]=row[1]
#caridadress

## Resolve and print output

In [19]:
#JUST WU
uni = '1545931'
print(uniname.get(uni))
for k, v in disto[uni].items():      
    if k in caridadress.keys():
        id = caridadress.get(k)
        type = 'Car'
    if k in bikeidadress.keys():
        id = bikeidadress.get(k)
        type = 'Bike'

    if k[0:3] != 'st_':
        print(type +"\t"+ str(v*1000)[0:5] +" m  \t" +id)

Wirtschaftsuniversität Wien
Bike	474.2 m  	Messeplatz
Bike	614.1 m  	Krieau
Car	526.3 m  	Vorgartenstraße 223
Bike	167.9 m  	Südportalstraße


## And now for all of em...

In [78]:
for unikey in uniname.keys():
    print(uniname.get(unikey) +"\n" + "-"*50)
    if (int(disto[unikey].get('st_car')) or int(disto[unikey].get('st_bike'))) > 1:
        print( "Total Cars: " + str(disto[unikey].get('st_car')) +"\t\tAverage Distance: "+str(disto[unikey].get('st_caravg'))[0:5]+" km")
        print( "Total Bikes: " + str(disto[unikey].get('st_bike')) +"\t\tAverage Distance: "+str(disto[unikey].get('st_bikeavg'))[0:5]+" km")
        print("-"*50)
    for k, v in disto[unikey].items():      
        if k in caridadress.keys():
            id = caridadress.get(k)
            type = 'Car'
        if k in bikeidadress.keys():
            id = bikeidadress.get(k)
            type = 'Bike'

        if k[0:3] != 'st_':
            print(type +"\t"+ str(v*1000)[0:5] +" m  \t" +id)
    print("\n")

Universität für Bodenkultur Wien
--------------------------------------------------
Bike	897.2 m  	Aumannplatz
Car	918.7 m  	Währinger Straße ggü. 149


Universität für Musik und darstellende Kunst Wien
--------------------------------------------------
Total Cars: 3		Average Distance: 0.574 km
Total Bikes: 7		Average Distance: 0.752 km
--------------------------------------------------
Bike	548.1 m  	Bahngasse
Bike	626.2 m  	Schwarzenbergplatz
Bike	630.2 m  	Arbeiterkammer
Bike	998.6 m  	Wien Mitte
Bike	557.0 m  	Fasanplatz
Bike	921.4 m  	Johannesgasse
Car	826.0 m  	Theresianumgasse
Car	303.3 m  	Reisnerstraße 35
Car	592.5 m  	Zaunergasse 1-3
Bike	987.4 m  	Sankt-Elisabeth-Platz


Webster University Vienna
--------------------------------------------------


Campus der Universität Wien
--------------------------------------------------
Total Cars: 8		Average Distance: 0.758 km
Total Bikes: 6		Average Distance: 0.625 km
--------------------------------------------------
Car	145.2 m  	A

# Open Todo for future work

## Add further data sets
- Subway stations, bus stations,
- parking garages and number of spots
- regular parking spaces and bike posts

## Categorize Overlaps (?) 

## Store final results in a sensible manner